### PR TITLE
Multiple repositories

### PR DESCRIPTION
**Pull request by weaverryan**
Combination of #23 and #31

### CODE REVIEW

**Comment by HeahDude** [comment]
Looks good! Thanks!

File: src/AppBundle/Issues/NullStatusApi.php [hunk 1]
```diff
@@ -1,21 +1,22 @@
 <?php
 
 namespace AppBundle\Issues;
+use AppBundle\Repository\Repository;
```
Comment (Ener-Getick) [inline src/AppBundle/Issues/NullStatusApi.php]: Needs a new line here

File: src/AppBundle/Issues/StatusApi.php [hunk 2]
```diff
@@ -1,6 +1,7 @@
 <?php
 
 namespace AppBundle\Issues;
+use AppBundle\Repository\Repository;
```
Comment (Ener-Getick) [inline src/AppBundle/Issues/StatusApi.php]: same here

File: src/AppBundle/Repository/Provider/InMemoryRepositoryProvider.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+            if (!isset($repositoryData['subscribers'])) {
+                throw new \InvalidArgumentException(sprintf('The repository %s is missing a subscribers key!', $repositoryFullName));
+            }
+
+            if (empty($repositoryData['subscribers'])) {
+                throw new \InvalidArgumentException(sprintf('The repository "%s" has no subscribers configured.', $repositoryFullName));
+            }
+
+            $this->addRepository(new Repository(
```
Comment (Ener-Getick) [inline src/AppBundle/Repository/Provider/InMemoryRepositoryProvider.php]: Shouldn't we add the secret key ?

Comment (Ener-Getick) [inline src/AppBundle/Repository/Provider/InMemoryRepositoryProvider.php]: ok sorry there's a ``@todo`` comment later

File: src/AppBundle/Tests/Issues/GitHub/CachedLabelsApiTest.php [hunk 4]
```diff
@@ -29,7 +29,7 @@ protected function setUp()
         $this->backendApi = $this->getMockBuilder('Github\Api\Issue\Labels')
             ->disableOriginalConstructor()
             ->getMock();
-        $this->api = new CachedLabelsApi($this->backendApi, self::USER_NAME, self::REPO_NAME);
+        $this->api = new CachedLabelsApi($this->backendApi, self::USER_NAME.'/'.self::REPO_NAME);
```
Comment (Ener-Getick) [inline src/AppBundle/Tests/Issues/GitHub/CachedLabelsApiTest.php]: the second parameter is useless now

File: src/AppBundle/Tests/Issues/GitHub/GitHubStatusApiTest.php [hunk 5]
```diff
@@ -30,7 +30,7 @@ protected function setUp()
         $this->labelsApi = $this->getMockBuilder('AppBundle\Issues\GitHub\CachedLabelsApi')
             ->disableOriginalConstructor()
             ->getMock();
-        $this->api = new GitHubStatusApi($this->labelsApi, self::USER_NAME, self::REPO_NAME);
+        $this->api = new GitHubStatusApi($this->labelsApi, self::USER_NAME.'/'.self::REPO_NAME);
```
Comment (Ener-Getick) [inline src/AppBundle/Tests/Issues/GitHub/GitHubStatusApiTest.php]: same

**Comment by Ener-Getick** [comment]
Loogs good as far as I can see :+1: 
Do the tests pass ?

**Comment by weaverryan** [comment]
@Ener-Getick the tests just passed after the last commit! :)

File: src/AppBundle/Repository/Repository.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
+<?php
+
+namespace AppBundle\Repository;
+use AppBundle\Issues\GitHub\GitHubStatusApi;
```
Comment (Ener-Getick) [inline src/AppBundle/Repository/Repository.php]: missing space :P (maybe you should run php-cs-fixer ?)

File: src/AppBundle/Subscriber/BugLabelNewIssueSubscriber.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+     * Changes "Bug" issues to "Needs Review".
+     *
+     * @param GitHubEvent $event
+     */
+    public function onIssues(GitHubEvent $event)
+    {
+        $data = $event->getData();
+        $repository = $event->getRepository();
+        if ('labeled' !== $action = $data['action']) {
+            $event->setResponseData(array('unsupported_action' => $action));
```
Comment (Ener-Getick) [inline src/AppBundle/Subscriber/BugLabelNewIssueSubscriber.php]: this should be removed imo as another listener could support this action

File: src/AppBundle/Subscriber/NeedsReviewNewPRSubscriber.php [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        ));
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return array(
+            GitHubEvents::PULL_REQUEST => 'onPullRequest',
+        );
+    }
+}
```
Comment (Ener-Getick) [inline src/AppBundle/Subscriber/NeedsReviewNewPRSubscriber.php]: Eof missing

File: src/AppBundle/Subscriber/BugLabelNewIssueSubscriber.php (hunk 7 above)
Comment (weaverryan) [inline src/AppBundle/Subscriber/BugLabelNewIssueSubscriber.php]: Yea, we need to work on this part a bit more - different listeners will replace each other's response data keys. Mostly, I just want a simple summary of what happened in the response, so we can use that in the functional tests. I'm not going to change this now, but yea, it's shaky.

File: src/AppBundle/Subscriber/NeedsReviewNewPRSubscriber.php (hunk 8 above)
Comment (weaverryan) [inline src/AppBundle/Subscriber/NeedsReviewNewPRSubscriber.php]: thx - I finally stopped being lazy and properly ran the CS fixer just now :)

**Comment by Ener-Getick** [comment]
Great thanks ! :smile:

**Comment by weaverryan** [comment]
Thanks for the quick review on this guys! It was great building off of both of your ideas really :).

Now, we can move on from here, adding new individual items :)

**Comment by HeahDude** [comment]
Thanks guys!

File: src/AppBundle/Issues/GitHubRequestHandler.php [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        return $responseData;
+    }
+
+    private function authenticate($hash, $key, $data)
+    {
+        if (!extension_loaded('hash')) {
+            throw new \RuntimeException('"hash" extension is needed to check request signature.');
+        }
+
+        return $hash !== 'sha1='.hash_hmac('sha1', $data, $key);
```
Comment (sstok) [inline src/AppBundle/Issues/GitHubRequestHandler.php]: The GitHub docs recommend to use a constant timing comparison.

File: src/AppBundle/Repository/Repository.php [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+     */
+    public function getVendor()
+    {
+        return $this->vendor;
+    }
+
+    /**
+     * @param string $vendor
+     */
+    public function setVendor($vendor)
```
Comment (sstok) [inline src/AppBundle/Repository/Repository.php]: Should the repository class be mutable?

File: src/AppBundle/Issues/GitHubRequestHandler.php (hunk 9 above)
Comment (weaverryan) [inline src/AppBundle/Issues/GitHubRequestHandler.php]: Can you open a PR?

File: src/AppBundle/Repository/Repository.php (hunk 10 above)
Comment (weaverryan) [inline src/AppBundle/Repository/Repository.php]: Ah, nice catch! Fixed at sha: 6f88a19e662f3ce5da01f6a3f49a14bec1f62aa9

File: src/AppBundle/Issues/GitHubRequestHandler.php (hunk 9 above)
Comment (sstok) [inline src/AppBundle/Issues/GitHubRequestHandler.php]: Will do 👍

Comment (sstok) [inline src/AppBundle/Issues/GitHubRequestHandler.php]: Done https://github.com/carsonbot/carsonbot/pull/40

**Pull request merged**